backend/src/interactions/updateUnconfirmedContribution/UnconfirmedContributionAdmin.role.ts

```diff
@@ -0,0 +1,56 @@
+import { Contribution } from '@entity/Contribution'
+import { User } from '@entity/User'
+
+import { RIGHTS } from '@/auth/RIGHTS'
+import { Role } from '@/auth/Role'
+import { AdminUpdateContributionArgs } from '@/graphql/arg/AdminUpdateContributionArgs'
+import { ContributionStatus } from '@/graphql/enum/ContributionStatus'
+import { LogError } from '@/server/LogError'
+
+import { UnconfirmedContributionRole } from './UnconfirmedContribution.role'
+
+export class UnconfirmedContributionAdminRole extends UnconfirmedContributionRole {
+  public constructor(
+    contribution: Contribution,
+    private updateData: AdminUpdateContributionArgs,
+    private moderator: User,
+  ) {
+    super(
+      contribution,
+      updateData.amount ?? contribution.amount,
+      updateData.creationDate ? new Date(updateData.creationDate) : contribution.contributionDate,
+    )
+  }
+
+  protected update(): void {
+    this.self.amount = this.updatedAmount
+    this.self.memo = this.updateData.memo ?? this.self.memo
+    this.self.contributionDate = this.updatedCreationDate
+    this.self.contributionStatus = ContributionStatus.PENDING
+    this.self.updatedAt = new Date()
+    this.self.updatedBy = this.moderator.id
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  protected checkAuthorization(user: User, role: Role): UnconfirmedContributionRole {
```

Comment:
Efficient in which sense?  
checkAuthorization will be called in the base class.
If I first have to check which derived class is currently active to decide which function to call, then I can do without the derivation. I don't see any use in dividing the action checkAuthorization further.

Reply:
As you said the derived class will invoke the `checkAuthorization` of the base-class.
So the derived class know itself and know if it has the User or Role data in its context. In consequence the invoker class knows directly if it can invoke the `checkAuthorization`with the parameter `User` or with the parameter `Role`.

I interpret the implementation of the base class method `checkAuthorization` with the first mandatory parameter `User` and the second optional parameter `Role`. But for the usage without `User` you need the special eslint-rule. That's the reason why i ask to devide the methode in the two specifications.

Reply:
in AbstractUnconfirmedContributionRole::checkAndUpdate 
checkAuthorization will be called. 
in context: pseudo code without additional parameter
```ts
let unconfirmedContributionRole: AbstractUnconfirmedContributionRole | null = null
if (admin user) {
  unconfirmedContributionRole = new UnconfirmedContributionAdminRole()
else if (user) {
  unconfirmedContributionRole = new UnconfirmedContributionUserRole()
}
unconfirmedContributionRole.checkAndUpdate(() => {
  this.checkAuthorization()
})
``` 
So it is the other way around. The Context which create the role, call the base class method checkAndUpdate which call the for AbstractUnconfirmedContributionRole  abstract function checkAuthorization which will have either the implementation from user or from admin. 

We are still using TypeScript and if a parameter in TypeScript is optional it will be marked with an ?
So for me checkAuthorization is a function with two mandatory parameters which will be booth filled from calling function, if they are needed in the end is up to the roles. Maybe we get more roles in the future, so it would be nice if it is possible to add a new role with at minimal code changes as possible in context.